### PR TITLE
GCN with MolGraphConv Featurization on BACE

### PR DESCRIPTION
**Pull request by mufeili**
@rbharath This PR is for a submission of GCN with MolGraphConv featurization on BACE. It includes:

- An entry for the submission in the leaderboard
- A script for training and hyperparameter search as well as a README file for the script. I want to reuse the script as much as possible across models and datasets.

It depends on [PR #2328](https://github.com/deepchem/deepchem/pull/2328) in DeepChem for a bug fix.

We also need to fix a warning related to DGL in DeepChem.

```bash
/home/ubuntu/anaconda3/envs/deepchem/lib/python3.6/site-packages/dgl/base.py:45: DGLWarning: DGLGraph.__len__ is deprecated.Please directly call DGLGraph.number_of_nodes.
  return warnings.warn(message, category=category, stacklevel=1)
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,6 +4,7 @@
...
-| 1    | Random Forest | 1024-bit ECFP2 | 0.8507 +- 0.0072 | 0.7368 +- 0.0066   | [Mufei Li](mufeili1996@gmail.com) | [Paper](https://www.stat.berkeley.edu/~breiman/randomforest2001.pdf), [Code](./examples) | Dec 2nd, 2020 |
+| Rank | Model         | Featurization  | Test ROC-AUC     | Validation ROC-AUC | Contact                           | References	                                                                           | Date           |
+| ---- | ------------- | -------------- | ---------------- | ------------------ | --------------------------------- | ---------------------------------------------------------------------------------------- | -------------- |
+| 1    | Random Forest | 1024-bit ECFP2 | 0.8507 +- 0.0072 | 0.7368 +- 0.0066   | [Mufei Li](mufeili1996@gmail.com) | [Paper](https://www.stat.berkeley.edu/~breiman/randomforest2001.pdf), [Code](./examples) | Dec 2nd, 2020  |
+| 2    | GCN           | GraphConv      | 0.8175 +- 0.0193 | 0.7430 +- 0.0194   | [Mufei Li](mufeili1996@gmail.com) | [Paper](https://arxiv.org/abs/1609.02907), [Code](./examples)                            | Dec 20th, 2020 |
```
Comment (rbharath) [inline README.md:10]: It's interesting the GCN underperforms the RF (but perhaps not that surprising since RFs often do very well)

File: examples/utils.py [hunk 2]
```diff
@@ -42,3 +50,50 @@ def init_trial_path(args):
...
+  if args['dataset'] == 'BACE':
+    tasks, all_dataset, transformers = load_bace_classification(
+        featurizer=featurizer, splitter=splitter, reload=False)
+  else:
+    raise ValueError('Unexpected dataset: {}'.format(args['dataset']))
+
+  return args, tasks, all_dataset, transformers
+
+
+class EarlyStopper():
```
Comment (rbharath) [inline examples/utils.py:75]: This is a useful utility class!

**Review by rbharath — APPROVED** [review]
This looks quite good! Feel free to merge in once ready :)

File: README.md (hunk 1 above)
Comment (mufeili) [reply README.md:10]: This can be a result of multiple reasons:

1. The featurization methods in the two submissions are different and the input features can contain different information.
2. The dataset is small. There is a significant performance gap between the validation set and the test set. In such cases, hand-crafted features can be more robust. Note that GCN does yield better validation performance as we only know validation performance in early stopping.

**Comment by mufeili** [comment]
@rbharath Thank you for the review. I think you can go ahead to merge it. Also copy @miaecle in case you have bandwidth to take a look.

**Comment by rbharath** [comment]
@mufeili The list of datasets on https://deepchem.readthedocs.io/en/latest/api_reference/moleculenet.html is probably pretty representative of MoleculeNet v2 (although we might have a few more datasets trickling in before we made the cut). We don't have a nice table with statistics though (which would be nice to have!)

Going to go ahead and merge this in (please still free to comment @miaecle if you have bandwidth!)
